internal/systemd/systemd.go

```diff
@@ -0,0 +1,174 @@
+package systemd
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	systemd "github.com/coreos/go-systemd/v22/dbus"
+	"github.com/subpop/go-log"
+)
+
+type ConnectionType int
+
+const (
+	ConnectionTypeSystem ConnectionType = iota
+	ConnectionTypeUser
+)
+
+type Conn struct {
+	ctx  context.Context
+	conn *systemd.Conn
+}
+
+// NewConnectionContext creates a new connection to the given systemd service.
+func NewConnectionContext(ctx context.Context, connectionType ConnectionType) (*Conn, error) {
+	var conn *systemd.Conn
+	var err error
+	if connectionType == ConnectionTypeSystem {
+		conn, err = systemd.NewSystemConnectionContext(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("cannot establish system connection to systemd: %v", err)
+		}
+	} else {
+		conn, err = systemd.NewUserConnectionContext(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("cannot establish user connection to systemd: %v", err)
+		}
+	}
+
+	return &Conn{
+		ctx:  ctx,
+		conn: conn,
+	}, nil
+}
+
+func (c *Conn) Close() {
+	c.conn.Close()
+}
+
+// Reload instructs systemd to scan for and reload unit files.
+func (c *Conn) Reload() error {
+	return c.conn.ReloadContext(c.ctx)
+}
+
+// EnableUnit enables the named unit. If activate is true, it also starts the
+// unit. If runtime is true, the unit is enabled for the runtime only (/run). If
+// false, it is enabled persistently (/etc).
+func (c *Conn) EnableUnit(name string, activate bool, runtime bool) error {
+	if _, _, err := c.conn.EnableUnitFilesContext(c.ctx, []string{name}, runtime, true); err != nil {
+		return fmt.Errorf("cannot enable unit %v: %v", name, err)
+	}
+
+	if activate {
+		if err := c.StartUnit(name, true); err != nil {
+			return fmt.Errorf("cannot start unit %v: %v", name, err)
+		}
+	}
+
+	return nil
+}
+
+// StartUnit starts the named unit. If wait is true, the method waits until the
+// unit state becomes "active".
+func (c *Conn) StartUnit(name string, wait bool) error {
+	done := make(chan string)
+	_, err := c.conn.StartUnitContext(c.ctx, name, "replace", done)
+	if err != nil {
+		return fmt.Errorf("cannot start unit %v: %v", name, err)
+	}
+	result := <-done
+	switch result {
+	case "done":
+		// The job successfully started, break to proceed
+		break
+	default:
+		return fmt.Errorf("failed to start unit with reason: %v", result)
+	}
+
+	if wait {
+		if err := c.waitForState(name, "active", 1*time.Second); err != nil {
+			return fmt.Errorf("timed out waiting for state 'active': %v", err)
+		}
+	}
+
+	return nil
+}
+
+// DisableUnit disables the named unit. If deactivate is true, it also stops the
+// unit. If runtime is true, the unit is enabled for the runtime only (/run). If
+// false, it is enabled persistently (/etc).
+func (c *Conn) DisableUnit(name string, deactivate bool, runtime bool) error {
```

Comment:
Similar comment here. What about renaming it to `DisableStopUnit()`?

Reply:
Yes, I had thought about `EnableStartUnit` and `DisableStopUnit`. This might be acceptable.